### DISCUSSION TITLE
3D Finite volume flow around a sphere

### DISCUSSION THREAD

**Discussion by adfboyd** [idaholab/moose#17470] [Q&A Modules: Navier-Stokes] [answered]
Hello, as a new user, I'm trying to work out how to simulate flow moving around a sphere, using finite volume incompressible flow in 3D.  I have looked at the examples for finite volume (channel flow especially), but how to convert to 3D and add the sphere object in is proving difficult. Any advice on where to look/what to do would be much appreciated!

**Comment by GiudGiud** [comment, marked as answer]
Hello

You will need meshing software to create your geometry as moose cant mesh the outside of a 3D sphere as far as I know. 
If you are a student / in academia you can use the student version of cubit. 

Once you have a mesh, setting up the problem in finite volume INS should not be too bad and we can help you with that too. 

Guillaume

**Reply by permcody** [reply]
I believe @dschwen has done this. He can probably provide a few pointers.

**Reply by permcody** [reply]
Here it is:
https://mooseframework.inl.gov/source/meshgenerators/SphereSurfaceMeshGenerator.html

I don't see a sample input file, but from the source it looks like the inputs are straightforward:
https://github.com/idaholab/moose/blob/f6e5d671f7597be193a65bab16cb0df270377497/modules/phase_field/src/meshgenerators/SphereSurfaceMeshGenerator.C#L29-L32

**Reply by adfboyd** [reply]
Thank you very much! I am a student, and this is very helpful! I may come back to ask about setting up the flow once I've tried it.

**Reply by adfboyd** [reply]
Now that the sphere can be generated, I'm afraid I'm still not sure how to set up a 3D flow and insert the sphere into it, are there any 3D examples for finite volume that I could look at? Thanks again!

**Reply by GiudGiud** [reply]
Are you using the SphereSurfaceMeshGenerator? Can you post a picture of the mesh?

There's no 3D examples but it should be very straightforward to extend 2D cases to 3D. Just add a w variable and use the same kernels.

**Reply by adfboyd** [reply]
I'm not sure how to use the SphereSurfaceMeshGenerator in order to actually generate an object, I'm afraid. I guess I'm also asking where I'm meant to do that. Is it done in the same application as the flow sim, or should it be used beforehand to make a file?
Thank you for your help again!

**Reply by GiudGiud** [reply]
It's a mesh generator. So you would use it in the mesh block, in the same input file as the flow sim.
I'm not convinced this is what you need though, but we ll see.

How about you start by creating a 3D channel with no sphere in it and we can advise you again once you have that working.

**Reply by adfboyd** [reply]
Thank you, that makes sense. I'll get that going.

**Reply by adfboyd** [reply]
I've edited the .i file for channel flow so that it should be suitable for 3D, however I'm afraid I'm not sure which kernels I should be linking to in order to run it. (I will test that the 2D example case runs first once I have this before seeking more help). Any help is much appreciated. Thanks

**Reply by GiudGiud** [reply]
I'm adding a 3D test to INSFV. It wont be merged very soon but you'll find it in #17088 in like an hour

**Reply by adfboyd** [reply]
Thank you for the 3D test, I understand how it was written a little better now. How can I actually get this to run? I've tried creating an application and copying in the test as the .i file, but am not sure how to correctly edit the Makefile. If this works, my other question is of course how I might add a fixed sphere into this simulation. Apologies if these questions are trivial, I can't seem to find answers in your tutorial.

**Reply by GiudGiud** [reply]
You need to add the navier Stokes module to your application in the Makefile. Just look for the lines where you choose modules. 

For the sphere you need to use a mesher to create it. 
No worries about the trivial questions, maybe they are in the tutorial, maybe not, just make sure to search the tutorial before posting.

**Reply by adfboyd** [reply]
Thank you. I did select the navier-stokes module before, my problem seems to be with library paths - on running I am told "A "INSFVelocityVariable" is not a registered object." I will look for solutions a bit more and possibly post in the getting started area if still stuck as that seems more appropriate. 
In the meantime I have downloaded Cubit as suggested and created a sphere. What file type should it be exported as? exodus?

**Reply by GiudGiud** [reply]
Exodus is the correct format for the mesh.
When you build your application do you see the stage where it's linking to the Navier Stokes library?
Is your moose up to date?

**Reply by adfboyd** [reply]
I can export the mesh as an exodus file then, great. Do I need to define boundaries in the meshing software, or can this be done in Moose?

I believe I did the first time, but now on building it says "Checking if header needs updating: .../moose/framework/include/base/MooseRevision.h...". I installed Moose from conda about 2 months ago, but haven't specifically checked for updates.

**Reply by GiudGiud** [reply]
It's a little easier to do in Cubit but you can do it in MOOSE. 
2 months should be fine for INSFV. That was merged before then

**Reply by GiudGiud** [reply]
can you report what `ldd your-application-opt` returns? (so ldd on your executable)

**Reply by adfboyd** [reply]
[ldd sphere_flow-opt.txt](https://github.com/idaholab/moose/files/6385401/ldd.sphere_flow-opt.txt)

This is a copy of the output from running the ldd command.

**Reply by adfboyd** [reply]
Would you recommend figuring out the BCs in cubit or Moose?

**Reply by GiudGiud** [reply]
You should tag/define the side set in Cubit, and use the sideset to define a BC in MOOSE.

Ok I see libnavier-stokes is linked. The problem might be that you mispelled INSF**VV**elocityVariable

**Reply by adfboyd** [reply]
Thank you, that fixed that problem. I thought I had just copy/pasted so didn't think to check that sort of error. I now get the error "Each subdomain must contain at least one kernel. The following block(s) lack an active kernel: 0"
I created a new .i file with the direct copy of the test file you uploaded, and am not sure where this came from.

**Reply by GiudGiud** [reply]
I ll check that input. It ran fine when I added it.

**Reply by adfboyd** [reply]
My previous version of the file made off editing the 2d version ran that far (it had other errors), but the line from it 
"kernel_coverage_check = false" seems to allow it to run fine.

**Reply by adfboyd** [reply]
I have successfully managed to run the channel flow case you wrote, and now have a (I think) appropriate sphere mesh from Cubit, with a defined sideset on the surface. How do I go about adding this into the channel flow simulation? Thanks again!

**Reply by ChrocheMisawa** [reply]
> I'm adding a 3D test to INSFV. It wont be merged very soon but you'll find it in #17088 in like an hour

I'm also working on a similar problem, except that I started with 2D. And I also need to include temperature and do a transient simulation.

I managed to use the FE method and was able to run successfully. However, the transient case for the FV method does not converge well. I know that the transient case for compressible is not fully supported but still worked fine for the first few time steps.

But I was not able to get the Incompressible to run, may be my BCs are not set up correctly, could you please show an example or see what is wrong with my input?

My input file used is attached.

[Flow_Past_a_Cylinder.zip](https://github.com/idaholab/moose/files/6433780/Flow_Past_a_Cylinder.zip)


Thank you!
Zeyu Chen

**Reply by GiudGiud** [reply]
@ChrocheMisawa please create a separate Discussions for your problem. 
@adfboyd So you have a mesh of the channel? Can you post a screenshot of your mesh if that's ok?
To use that mesh you'll need to use [FileMeshGenerator](https://mooseframework.inl.gov/source/meshgenerators/FileMeshGenerator.html) to load it

**Reply by adfboyd** [reply]
I have a mesh of the sphere, made in cubit - see the attached screenshot. I need to separately mesh the channel as well? I though that might be done in Moose. Thanks!
<img width="711" alt="Spherescrshot" src="https://user-images.githubusercontent.com/61464721/117316562-9db71380-ae80-11eb-9d69-64141d50a7c8.png">

**Reply by GiudGiud** [reply]
The flow will be going around the sphere. It's that part that needs to be meshed. Unless you are doing heat conduction in the sphere or something, the inside of the sphere does not need to be meshed.

There's no mesh generator to mesh around things in MOOSE. That is a very advanced mesh generation capability, which you'll find in Cubit.

**Reply by adfboyd** [reply]
Ah, ok, I misunderstood the goal, but what you're saying makes sense. And it will have to be a 3d mesh of the volume in the channel that isn't in the sphere? I'll have to look at that.

**Reply by GiudGiud** [reply]
Yes.
@ChrocheMisawa uploaded a 2D mesh that you can look at as an example.

**Reply by adfboyd** [reply]
Thank you for your help, I've been attempting to do this mesh in the meantime. 
A long term goal for my project is to simulate objects flowing through a fluid, with forces acting over the fluid-solid boundary and with 6DOF motion of the objects. (An example still obtained using an older solver is attached.) We were hoping that at some point this might be possible to do in MOOSE, given possibly a combination of your contact and NS modules, but your comment about meshing might suggest otherwise, given that this is a crucial step in solution. What are your thoughts on whether this would be possible?
![Symmertic_Far_field](https://user-images.githubusercontent.com/61464721/118277073-4b00db80-b4c0-11eb-9fca-b89e52b464be.png)

**Reply by GiudGiud** [reply]
so the 3D solid ball will be moving in the mesh?

**Reply by adfboyd** [reply]
That would be the goal, yes. This would also require the mesh to move each step to adapt to the moving solid.

**Reply by GiudGiud** [reply]
Hi, this can be done in moose. You should look at the arbitrary lagrangian eulerian capabilities (ALE).
The fluid mesh will need to be adapted on every time step, using a simulated diffusive or elastic process to make it match the outer surface of the ball. 
Is the ball rigid? If not you'll also need the fluid-structure interaction (FSI module) capabilities.

**Reply by adfboyd** [reply]
Hi, this sounds exactly like what we're hoping for. The ball/s (we hope to be able to study more than one solid) would be rigid. I will look into the ALE capabilities.

**Reply by GiudGiud** [reply]
@fdkong

**Reply by adfboyd** [reply]
I now have what I hope is a suitable mesh (see a version attached - I plan to use a much finer mesh). What are the next steps to adding it into the simulation? I have created 7 sidesets too - the size borders of the channel and the surface of the sphere.


[
<img width="1027" alt="VolMeshScrshot" src="https://user-images.githubusercontent.com/61464721/119377642-42c15100-bcb5-11eb-8410-c4aaf9c2683e.png">

**Reply by GiudGiud** [reply]
You can use a FileMeshGenerator to load this mesh into the simulation. 
https://mooseframework.inl.gov/source/meshgenerators/FileMeshGenerator.html
You should try to get a reasonable solution with the flow around a fixed ball first.

Since you are planning to use ALE, you should look at the finite element fluid flow capabilities. Those capabilities are not ready yet for finite volume. You can use the tests in modules/navier_stokes/test/tests/ins as examples.

**Reply by adfboyd** [reply]
Thanks, I will definitely stick to this problem for now.

**Reply by adfboyd** [reply]
I've been spending a long time trying to get MOOSE on a cluster, but I think that is mostly sorted, and I am back to trying to get this to run! 
I've attached the script I'm trying to use to do this, but I think I'm having a problem loading the Mesh - I get this error:
```
^[[31m
*** ERROR ***
Task init_mesh is not registered to build MeshGenerator derived objects^[[39m
```
Do you know what might be causing this?

[flowroundsphere.txt](https://github.com/idaholab/moose/files/6639719/flowroundsphere.txt)

**Reply by fdkong** [reply]
You need to change

```
[Mesh]
  type = FileMeshGenerator
  file = 'SphereChannelMeshM10.e'
[]
```

to 

```
[Mesh]
  [gmg]
    type = FileMeshGenerator
    file = 'SphereChannelMeshM10.e'
  []
[]
```

It is the syntax of mesh generator system

**Reply by adfboyd** [reply]
Thank you! I am now having this problem:
```
/lus/cls01095/work/e643/e643/shared/mooseprojects/sphere_flow/flowroundsphere.i:153: (FVBCs/walls-u/boundary):
    the following side set ids do not exist on the mesh: 8

    MOOSE distinguishes between "node sets" and "side sets" depending on whether
    you are using "Nodal" or "Integrated" BCs respectively. Node sets corresponding
    to your side sets are constructed for you by default.

    Try setting "Mesh/construct_side_list_from_node_list=true" if you see this error.
    Note: If you are running with adaptivity you should prefer using side sets.^[[39m
```
I have tried the included suggestion, it has no effect. I'm not sure why it wants 8 sideset IDs, I've included a picture of the mesh I'm using, and it does indeed have 7. 
<img width="933" alt="MeshedVol" src="https://user-images.githubusercontent.com/61464721/121726314-4856cd80-cae2-11eb-9c02-7c42635dc7b0.png">

**Reply by fdkong** [reply]
It is interesting. 

Run 

`./yourbinary -i flowroundsphere.i Mesh/construct_side_list_from_node_list=true --mesh-only`

And then use Paraview to check if the file  named "*in.e" has node sets

**Reply by adfboyd** [reply]
I now get this:
```
^[[31m
*** ERROR ***
The supplied name cannot be empty, to allow for any name to be supplied use the '*' character.^[[39m

MPICH ERROR [Rank 2] [job id 325248.0] [Fri Jun 11 18:38:06 2021] [unknown] [nid001024] - Abort(1) (rank 2 in$

aborting job:
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 2
```
and no outputted file looking like "*in.e"

**Reply by GiudGiud** [reply]
If you have 7 sidesets and it s looking for sideset 8, it means that it s looking for one that has not been defined
Please add `show_info = true` to your last mesh generator and report that

**Reply by adfboyd** [reply]
If you mean that I should add show_info=true into the .i file, in the [Mesh] section, I have done this and it has not changed the output. If I should do something else, please let me know!

**Reply by GiudGiud** [reply]
It has to be in the mesh generator block, not the [Mesh] one)
for example:
```
[Mesh]
  [gmg]
    type = FileMeshGenerator
    show_info = true
    ...
  []
[]
```

**Reply by adfboyd** [reply]
Results attached, thanks!
[moosetest-331137.txt](https://github.com/idaholab/moose/files/6657148/moosetest-331137.txt)

**Reply by GiudGiud** [reply]
The side set you are using in FVBCs/walls-u/boundary is not present in the mesh. You need to add it during your mesh generation or using mesh generators.

**Reply by fdkong** [reply]
Side sets might be fine, but the corresponding  node sets were not generated ... 

The corresponding node sets should be generated by default without user's effort.  This seems a bug somewhere

Where can I  download the mesh and input files? @adfboyd 

Sorry for inconvenience.

**Reply by adfboyd** [reply]
The sideset issue is solved - I noticed a slight spelling error in the sideset names. The input file is attached. The mesh file seems not to be supported here - you can find it at: https://drive.google.com/file/d/1IysWoTU3JqlM3WTJFUY1Ol7HTbfVseU4/view?usp=sharing
Apologies if that is not the best way to share it. 
[flowroundsphere.txt](https://github.com/idaholab/moose/files/6657272/flowroundsphere.txt)

**Reply by adfboyd** [reply]
I am now having a problem in that the solve doesn't seem to converge - is there something I should be changing in the solver settings to help this? I've attached the output. 
[moosetest-331199.txt](https://github.com/idaholab/moose/files/6662079/moosetest-331199.txt)

**Reply by GiudGiud** [reply]
The linear solve is failing. This usually means the problem is ill-posed or singular.
I don't see any problem with the mesh. 

These may not be the problem but a couple of things about your input file:
- you won't be able to initialize straight at viscosity of 1e-3. You ll need to ramp it down, using a GenericFunctionMaterial for example
- no need to comment out NEWTON as the solve type. It s the best one for anything using automatic differentiation, which FV is using
- you will not be able to do FSI with finite volume for the fluid flow. Not without extensive development. The finite volume capabilities are not compatible with everything yet.

**Reply by adfboyd** [reply]
It seems to work with viscosity of 1e-2. How do I use GenericFunctionMaterial to ramp it down? I'm afraid I don't really understand how it works. 
For the Newton solver, I thought that Transient and Newton were two different options, but I am glad that I am able to use both. In terms of FSI, that's alright. I may have been slightly mistaken earlier - should we use MOOSE (which is looking likely), we plan to spend several months developing the capability to use the immersed boundary method with finite volume fluid flow.

**Reply by GiudGiud** [reply]
So you create a function for the ramp which shows the value of velocity over time
```
[Functions]
  [mu_func]
    type = PiecewiseLinear
    x = '1 3 5 10'
    y = '1e3*${mu} 1e2*${mu} 1e1*${mu} *${mu}'
  []
[]
and then you use a material to define viscosity instead of just a single value

`
[Materials]
  [mu_mat]
    type = GenericFunctionMaterial
    prop_names = 'mu'
    prop_values = mu_func
  []
[]
`

Transient is an executioner 'type', so what problem it's trying to solve (transient, eigenvalue, steady state), while `Newton` is a solve type, what numerical method we should use. 

Ok that's great news! We should be able to help you a bit with that.

**Reply by adfboyd** [reply]
I'm not sure if I should be asking this here or in a separate post, but I'm trying to run this on an HPC cluster, and when using a large number of nodes, the problem seems to never initialise properly. I was wondering if there are any built-in profiling tools with MOOSE so that I could check where the time is being used on larger problems.

**Reply by GiudGiud** [reply]
Like it takes forever to initialize?
It might be the creation of the exodus, this format does not suit itself to large runs. You could try disabling the exodus formation and see if it helps. The parallel format is Nemesis
https://mooseframework.inl.gov/syntax/Outputs/index.html

**Reply by adfboyd** [reply]
Yes, I can wait 20 minutes and it won't even output the mesh info. So to change the output, I should change 
```
[Outputs]
  exodus = true
[]
```
to 
```
[Outputs]
  nemesis = true
[]
```
or 
```
[Outputs]
  [other]
    type = Nemesis
  []
[]
```?

**Reply by fdkong** [reply]
> [Outputs]
  nemesis = true
[]

should work. 

BTW, if you want to profile code in the future, this is a good tool we often use.https://mooseframework.inl.gov/application_development/profiling.html

For the current case, if the mesh is large, exodus may be slow since we need to gather everything to the first core.

**Reply by adfboyd** [reply]
Unfortunately just this change doesn't seem to have helped - is there anything else that I might do? Thanks!

**Reply by fdkong** [reply]
Could you post the entire screen prints? I was curious how much time we spend on each section.

**Reply by adfboyd** [reply]
If you mean the screen prints without the profiling, I'm afraid there is absolutely nothing! I'm trying to reinstall with the gperftools at the moment.

**Reply by GiudGiud** [reply]
You have to a add a PerfGraphOutput to the output section to get some screen prints with timings

**Reply by GiudGiud** [reply]
Any news on this @adfboyd

**Reply by adfboyd** [reply]
I'm afraid I got stuck on other projects for a bit - I'll let you know how it goes when I manage to try again, probably early next week!
